src/compression.rs

```diff
@@ -16,7 +27,354 @@ pub struct MessageChunk(u64);
+                    .try_into()
+                    .unwrap()
+            })
+    }
+
+    fn spread_c(&self) -> Value<[bool; 28]> {
+        self.c.spread.value().map(|v| v.0)
+    }
+
+    fn spread_d(&self) -> Value<[bool; 14]> {
+        self.d.spread.value().map(|v| v.0)
+    }
+}
+
+#[derive(Clone, Debug)]
+pub struct RoundWordDense(AssignedBits<16>, AssignedBits<16>);
+
+impl From<(AssignedBits<16>, AssignedBits<16>)> for RoundWordDense {
+    fn from(halves: (AssignedBits<16>, AssignedBits<16>)) -> Self {
+        Self(halves.0, halves.1)
+    }
+}
+
+impl RoundWordDense {
+    pub fn value(&self) -> Value<u32> {
+        self.0
+            .value_u16()
+            .zip(self.1.value_u16())
+            .map(|(lo, hi)| lo as u32 + (1 << 16) * hi as u32)
+    }
+}
+
+#[derive(Clone, Debug)]
+pub struct RoundWordSpread(AssignedBits<32>, AssignedBits<32>);
+
+impl From<(AssignedBits<32>, AssignedBits<32>)> for RoundWordSpread {
+    fn from(halves: (AssignedBits<32>, AssignedBits<32>)) -> Self {
+        Self(halves.0, halves.1)
+    }
+}
+
+impl RoundWordSpread {
+    pub fn value(&self) -> Value<u64> {
+        self.0
+            .value_u32()
+            .zip(self.1.value_u32())
+            .map(|(lo, hi)| lo as u64 + (1 << 32) * hi as u64)
+    }
+}
+
+#[derive(Clone, Debug)]
+pub struct RoundWordA {
+    pieces: Option<AbcdVar>,
+    dense_halves: RoundWordDense,
+    spread_halves: Option<RoundWordSpread>,
+}
+
+impl RoundWordA {
+    pub fn new(
+        pieces: AbcdVar,
+        dense_halves: RoundWordDense,
+        spread_halves: RoundWordSpread,
+    ) -> Self {
+        RoundWordA {
+            pieces: Some(pieces),
+            dense_halves,
+            spread_halves: Some(spread_halves),
+        }
+    }
+
+    pub fn new_dense(dense_halves: RoundWordDense) -> Self {
+        RoundWordA {
+            pieces: None,
+            dense_halves,
+            spread_halves: None,
+        }
+    }
+}
+
+#[derive(Clone, Debug)]
+pub struct RoundWordE {
+    pieces: Option<EfghVar>,
+    dense_halves: RoundWordDense,
+    spread_halves: Option<RoundWordSpread>,
+}
+
+impl RoundWordE {
+    pub fn new(
+        pieces: EfghVar,
+        dense_halves: RoundWordDense,
+        spread_halves: RoundWordSpread,
+    ) -> Self {
+        RoundWordE {
+            pieces: Some(pieces),
+            dense_halves,
+            spread_halves: Some(spread_halves),
+        }
+    }
+
+    pub fn new_dense(dense_halves: RoundWordDense) -> Self {
+        RoundWordE {
+            pieces: None,
+            dense_halves,
+            spread_halves: None,
+        }
+    }
+}
+
+#[derive(Clone, Debug)]
+pub struct RoundWord {
+    dense_halves: RoundWordDense,
+    spread_halves: RoundWordSpread,
+}
+
+impl RoundWord {
+    pub fn new(dense_halves: RoundWordDense, spread_halves: RoundWordSpread) -> Self {
+        RoundWord {
+            dense_halves,
+            spread_halves,
+        }
+    }
+}
```

Comment:
We don't need these exact changes for BLAKE2, I don't think we'll need different structs for the round words because we'll be able to make them all the same size.

Reply:
Ill start removing them.  Devils advocate, if we do move on sha2-256 this maybe useful to keep, it doesnt break blake2, that being said it maybe better to have different ones for optimization purposes. Either way ill move to remove it, your more experienced in the subject so put way more weight towards you

Reply:
Yeah, I do agree if we were doing SHA it would be ideal! Let's discuss if we want to do both.